+ Security for Metrics Endpoint

becky-intelletive opened the pull request:
- Env vars added:
```
METRICS_DISABLE_ENDPOINT  # default: false
METRICS_SECURE_ENDPOINT  # default: false
METRICS_API_KEY  # default: false
METRICS_MINIMUM_AUTHORITY_LEVEL  # default: false
```
- Add API_KEY authority (including db migration)
- Add metrics to config service
- Add metrics guard
   - throw 404 if endpoint is disabled
   - if endpoint is secured,
      - if there's a logged-in user, check if they have the minimum required authority
      - if not, try to retrieve the api key and use the key to retrieve the user
         - if there's a user, use that to check if they have the minimum required authority
         - if there's no user, check if the api key is the same as the configured allowed key
      - checking the minimum required authority checks if the user has the authority defined in METRICS_MINIMUM_AUTHORITY_LEVEL, a greater authority, or the API_KEY authority
   - if endpoint is NOT secured, allow access
- `api-key.dao.ts`: add dao function to check if the api key exists
- Command module:
   - Add metrics key to the helm setup command
   - Add method & command to create an api key
- `LoggedInUserFactory.ts`: fix spacing (no other changes)

kyle-intelletive commented on dash/backend/src/config/metrics.ts:
```diff
@@ -0,0 +1,8 @@
+export default {
+  disableEndpoint: (process.env.METRICS_DISABLE_ENDPOINT || 'false').toString().toLowerCase() === 'true',
+  secureEndpoint: (process.env.METRICS_SECURE_ENDPOINT || 'false').toString().toLowerCase() === 'true',
```
The rest of the config uses 1 or 0 to determine boolean config, this should be made consistent.

See database.migrations.enabled, email.system.enableErrorReportEmail, rabbitMQ.enabled for patterns (Annoyingly, they are all slightly different)

kyle-intelletive commented on dash/backend/src/modules/command-line/commands/helm-setup.command.ts:
```diff
@@ -14,11 +14,12 @@ import {generateRandomString} from "./generate-random-string";
  */
 @Injectable()
 export class HelmSetupCommand {
+    promises: Promise<any>[] = [];
```
It doesn't look you're actually using this new variable

kyle-intelletive commented on dash/backend/src/modules/command-line/services/cli-command-builder.service.ts:
```diff
@@ -61,6 +61,12 @@ export class CliCommandBuilderService {
                 'Accepts comma delimited list of namespaces to whitelist',
                 () => this.helmSetupCommand.loadDefaultNamespaceExceptions(),
             ),
+            new CliCommand(
```
Why do we need a CLI command for this? We have a UI for creating API Keys.

becky-intelletive replied:
Eh, fair enough. Though we could ask the same question of the runSeed() job.

kyle-intelletive commented on dash/backend/src/modules/user/enum/Authority.ts:
```diff
@@ -4,6 +4,7 @@ export enum Authority {
-    FALCO = 'FALCO'
+    TRAWLER = 'TRAWLER',
+    FALCO = 'FALCO',
+    API_KEY = 'API_KEY',
```
This authority level should probably be called 'Metrics' or something similar.

It is very strange to have most API Keys not have the API Key level, and the Authority Level API KEY granting you access to the metrics endpoint is not intuitive.

becky-intelletive replied:
We don't really have anywhere else we use api keys. In my mind, this was a way to start granting them access to things.

The UI attaches API Keys to specific user profiles. This approach is more of a service-based API Key: giving other programs access to endpoints.

becky-intelletive replied:
I can change it so it's more specific. I just wanted to start making a path forward for programmatic access to endpoints.

becky-intelletive replied:
Instead of having to make an endless number of Authorities based on specific criteria ;)

kyle-intelletive reviewed (changes requested):
2 overall comments
1. It looks like you missed committing your migration
2. Is there a reason you are making these API Keys function so differently from the other ones?

becky-intelletive commented:
@kyle-intelletive -
1. I didn't commit it because I didn't want to add to the migration tree yet.
2. See the comment in the guard. I think we could move that functionality into the logged-in user population file. It would allow us to pass in API keys as bearer tokens or headers. Beyond that, it's another programmatic API Key. Those are handled differently than users' individual API keys (which are attached to their permissions). I want to open access to the endpoint to different classes of "users":
   - ones with defined access levels who are logged in
   - ones with defined access levels who are not logged in but are using individual API keys (granted via the ui)
   - services that are using their assigned API key
   - services that are using the global API key

becky-intelletive commented:
@kyle-intelletive 
- Moved more validation from the guard to the LOGGED_IN_USER provider
   - LOGGED_IN_USER provider can now get api keys from bearer tokens and headers
- Authority.METRICS instead of Authority.API_KEY
- Checked in the migration
- Added a note to "save kubehunter report" to utilize the LOGGED_IN_USER provider

Any more thoughts on how it can be simplified/genericized?

kyle-intelletive commented:
> @kyle-intelletive
> 

>       
>       * LOGGED_IN_USER provider can now get api keys from bearer tokens and headers
> 
>     * Authority.METRICS instead of Authority.API_KEY
> 
>     * Checked in the migration
> 
>     * Added a note to "save kubehunter report" to utilize the LOGGED_IN_USER provider
> 
> 
> Any more thoughts on how it can be simplified/genericized?
Given that the LOGGED_IN_USER is factory, it will be kind of difficult 

I can think of a couple ways. Here are some quick overviews of them, I can discuss them in more detail at some point if you'd like.

# A. Change to service
If we use a service rather than factory (still scoped to request) - we could implement a decorator like below that would basically just call that service with the appropriate config. then we'd change our references to call this.LOGGED_IN_USER.get() rather than it being a defined variable. We'd also have to ensure that the Authentication() decorator is resolved before any other decorators that care about the user 

# B. Make mono-decorator
Basically, we roll the functionality of UseGaurds, AllowedAuthorityLevels, and LOGGED_IN_USER into one decorator.
then, we'd also have to define another decorator that functions like the @Body() decorator, that pulls the value of the user from the metadata that the main decorator will call. (Won't need that secondary decorator if we only care that the user is authenticated and not who they are)

# C. Provide variations of the LOGGED_IN_USER
We can inject variations of the LOGGED_IN_USER factory that handle auth differently. Off the top of my head, I know we can do this at a module level, we'd have to do more research to see if it can be changed at the class/function level 


# Sample code for both A & B
```
@Authentication({
 authorityLevels: [Authority.Admin ...]
  // If true, checks for bearer token as "normal user"
  allowUsers: true, 
  // If true, checks for API Key users
  allowApiKeys: true, 
  // optional, checks some reasonable default if not set
  apiKeySources: {
     // Checks in the headers and query params of provided names
     headers: ['x-api-key', 'whatever-else' .....]
     queryParams: ['key', 'apiKey' ....],
     // Check if bearer token is actually an api key
     bearerToken: true
  }
})
myEndpoint() {
...
}
```

becky-intelletive commented:
> > @kyle-intelletive
> I can think of a couple ways. Here are some quick overviews of them, I can discuss them in more detail at some point if you'd like.
> ...
That seems like an expansion of scope to improved the way that the app fills/handles LOGGED_IN_USER (and subsequent authorization). Do you think this would make more sense as a new ticket?

kyle-intelletive commented:
> > > @kyle-intelletive
> > > I can think of a couple ways. Here are some quick overviews of them, I can discuss them in more detail at some point if you'd like.
> > > ...
> > > That seems like an expansion of scope to improved the way that the app fills/handles LOGGED_IN_USER (and subsequent authorization). Do you think this would make more sense as a new ticket?

@becky-intelletive 
For sure, this would definitely be a different feature. I will review the changes made from the other topics shortly.

becky-intelletive commented:
> @becky-intelletive For sure, this would definitely be a different feature. I will review the changes made from the other topics shortly.

https://dev.azure.com/intelletive/Minesweeper/_boards/board/t/Minesweeper%20Team/Issues/?workitem=6897

kyle-intelletive commented on dash/backend/src/modules/kube-hunter/dto/kube-hunter-raw.interface.ts:
```diff
@@ -0,0 +1,90 @@
+  exposeDefaultValues: boolean
+  exposeUnsetFields: boolean
+  ignoreDecorators: boolean
+}
```
A lot of the interfaces in this file are very generically named. It would probably be a good idea to name them something like KubeHunterWhatever.

Right now all the generic ones are just used in IKubeHunterRawResponseSection, but in the future, if we have a reason to do more processing on the reports, we may need to export these inner interface

becky-intelletive replied:
So I generated the interfaces by throwing the response into an interface generator ;)
I had made the names slightly more explicit and consolidated repeat interfaces, but I hadn't bothered making them too clear since they were internal interfaces.
The latest commit adds `IKHRawResponse` to the beginning of each interface.

kyle-intelletive commented on manifests/charts/dash/templates/prometheus-rules-default.yaml:
```diff
@@ -17,7 +17,7 @@ spec:
       - alert: ManyNewCVEs
         annotations:
-          description: Fires when the number of new major + critical CVEs in the last 10m exceeds 20
+          description: Fires when the number of new major + critical CVEs in the last 10m exceeds {{ .threshold }}
```
Should this have use a .duration like the ones below?

becky-intelletive replied:
The .duration field is actually new

becky-intelletive commented:
@kyle-intelletive - ready for re-review
